### DISCUSSION TITLE
Imports needed to edit custom scripts in vscode

### DISCUSSION THREAD

**Discussion by tux-mania** [netbox-community/netbox#9399] [Q&A] [answered]
Hi all,

I am trying to edit custom scripts with VSCode.
Thanks to @drmsoffall ([post here](https://github.com/netbox-community/netbox/discussions/9380)), I know now how to get objects properties with `./manage.py nbshell`.
In order to manage the code in an easy way, I prefer to connect remotely with SSH (using VSCode plugin) and edit the scripts. The thing is: I don't know which libraries need to be import to have the objects available (and therefor get the autocompletion and help when coding with VSCode).

Is there a way to do this?

Thanks for your help!

**Comment by mtinberg** [comment, marked as answer]
If you open the editor at the top of the directory structure for netbox where the venv is defined, can you select the virtualenv python as your python interpreter?  Having VSCode using the same python virtualenv as the application will mean it can resolve the same libraries, just activating a virtualenv in a terminal window inside of VSCode will only activate the virtualenv in that shell instance, not for the whole VSCode process and not for the python language support tools which need to resolve the right library paths to work.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: dsc999 ***@***.***>
Sent: Friday, May 20, 2022 5:05 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Imports needed to edit custom scripts in vscode (Discussion #9399)


Hi all,

I am trying to edit custom scripts with VSCode.
Thanks to @drmsoffall<https://github.com/drmsoffall> (post here<https://github.com/netbox-community/netbox/discussions/9380>), I know now how to get objects properties with ./manage.py nbshell.
In order to manage the code in an easy way, I prefer to connect remotely with SSH (using VSCode plugin) and edit the scripts. The thing is: I don't know which libraries need to be import to have the objects available (and therefor get the autocompletion and help when coding with VSCode).

Is there a way to do this?

Thanks for your help!

**Reply by tux-mania** [reply]
@mtinberg thank you for your help!
Actually, dumb question and nothing to do with netbox but with VS Code use (sorry about that)... 
Works perfectly this way. Thanks again!

**Reply by llamafilm** [reply]
Have you found a way to make VS Code recognize Netbox internal modules?  I set it to use the Python interpreter inside the venv (`/opt/netbox/venv/bin/python`) but it still doesn't recognize internal modules so I get the yellow warning lines like this:

<img width="415" alt="image" src="https://github.com/netbox-community/netbox/assets/3991046/5e211a2b-316a-4694-96f9-8e63269d80dc">

**Reply by llamafilm** [reply]
Nevermind, I figured it out!  I added to VS Code settings.json: `{"python.envFile": "${workspaceFolder}/.vscode/.env"}`

And added `PYTHONPATH=/opt/netbox/netbox` to that file.

**Reply by pgnuta** [reply]
@llamafilm... This!  I've been looking for how to fix this forever!
